### PR TITLE
fix: Include md5 hash of checksum when checksum isn't a valid hash on…

### PR DESCRIPTION
**Pull request by bretthoerner**
… its own

Some users send checksums that aren't hash-like at all, which makes it
very confusing for downstream consumers like Snuba. This includes a
proper md5 sum as the 'primary hash' if the user's checksum
doesn't already appear to be a valid hash.

---

Not sure what exactly this means for migrating existing data. Any group that receives a new event should get a proper md5 attached to it so that future Snuba queries work. I don't forsee any bad effects on existing data, though? Need a ted here.

I included the same code on the Snuba writer side so that we're backwards compatible with old data in Kafka: https://github.com/getsentry/snuba/pull/3

### CODE REVIEW

File: src/sentry/event_manager.py [hunk 1]
```diff
@@ -46,6 +47,7 @@
 from sentry.stacktraces import normalize_in_app
 
 
+HASH_RE = re.compile(r'^[0-9a-f]{32}$')
```
Comment (alex-hofsteede) [inline src/sentry/event_manager.py]: case insensitive?

File: src/sentry/event_manager.py [hunk 2]
```diff
@@ -600,7 +602,10 @@ def save(self, project, raw=False):
         if fingerprint:
             hashes = [md5_from_hash(h) for h in get_hashes_from_fingerprint(event, fingerprint)]
```
Comment (alex-hofsteede) [inline src/sentry/event_manager.py]: not so much related to this PR, but just a clarification on the snuba stuff, if primary_hash is just the hash of the first element, does that mean that an event with fingerprint `['foo', 'bar']` has the same primary hash as one with a fingerprint of `['foo', 'baz']` and is that the correct matching behavior we want?

**Review by alex-hofsteede — APPROVED** [review]

File: src/sentry/event_manager.py (hunk 1 above)
Comment (bretthoerner) [reply src/sentry/event_manager.py]: Good call, is JSON-schema always case insensitive? If not I think we need to adjust the Snuba schema too.

File: src/sentry/event_manager.py (hunk 2 above)
Comment (bretthoerner) [reply src/sentry/event_manager.py]: Stream of consciousness here:

* an event can have multiple hashes/fingerprints: `[a, b, c]`
* hexdigest hashes are calculated in `event_manager.save`
* each one is used in a `GroupHash.objects.get_or_create(project, hash)`
* those are iterated over and the first `GroupHash` that already has a `group_id` is used as the group, else a new group is created
* that `group_id` is associated with *all* of the `GroupHash` values from earlier (representing `[a, b, c]`)

So we use the first hash at write time in Snuba... and at query time (when searching a group) we will always send all hashes associated with the group.

That was my brain dump... if you see any holes now is the time for us to change stuff. :fearful:

Comment (bretthoerner) [reply src/sentry/event_manager.py]: I guess there is some fuzziness around merge and unmerge and what it means when the `primary_hash` for a given set of events is moved between groups. @tkaemming would probably have a better idea here.

**Comment by tkaemming** [comment]
> Any group that receives a new event should get a proper md5 attached to it so that future Snuba queries work. I don't forsee any bad effects on existing data, though? Need a ted here.

Seems good to me for data moving forward. 👍

File: src/sentry/event_manager.py (hunk 2 above)
Comment (tkaemming) [reply src/sentry/event_manager.py]: Nothing really significant with merging will change here. Unmerging will be a little weird (you'd have to move two `GroupHash` records when previously you only had to move one) but I don't think that's a significant enough problem to worry about based on the low utilization of `checksum` anyway.

**Pull request merged**